bulletroboy/state_mapper/state_mapper.py

```diff
@@ -0,0 +1,382 @@
+	def __init__(self):
+		super().__init__('state_mapper')
+		self.declare_parameters(
+			namespace='',
+			parameters=[
+				('roboy_link_names', None),
+				('operator_link_names', None)
+			])
+		
+		self.callback_group = ReentrantCallbackGroup()
+
+		self.roboy_link_names = self.get_parameter("roboy_link_names").get_parameter_value().string_array_value
+		self.operator_link_names = self.get_parameter("operator_link_names").get_parameter_value().string_array_value
+		
+		assert len(self.roboy_link_names) == len(self.operator_link_names), "Roboy and operator links list have different lenghts"
+
+		self.roboy_initial_link_poses = {}
+		self.operator_initial_link_poses = {}
+		
+		#to avoid deadlocks in services
+		self.roboy_is_ready = False
+
+		# Define clients
+		self.roboy_link_info_from_id_client = self.create_client(LinkInfoFromId, Services.LINK_INFO_FROM_ID, callback_group=self.callback_group)
+		self.operator_link_info_from_name_client = self.create_client(LinkInfoFromName, Services.LINK_INFO_FROM_NAME, callback_group=self.callback_group)
+		self.roboy_initial_link_pose_client = self.create_client(GetLinkPose, Services.ROBOY_INITIAL_LINK_POSE, callback_group=self.callback_group)
+		self.operator_initial_link_pose_client = self.create_client(GetLinkPose, Services.OP_INITIAL_LINK_POSE, callback_group=self.callback_group)
+
+		# Define subscriptions
+		self.robot_collision_subscription = self.create_subscription(
+			Collision,
+			Topics.ROBOY_COLLISIONS,
+			self.collision_listener,
+			1)
+		# Operator EF pose subscriber
+		self.operator_movement_subscription = self.create_subscription(
+			PoseStamped, 
+			Topics.OP_EF_POSES, 
+			self.operator_movement_listener, 
+			1)
+		# Define publishers
+		self.exoforce_collision_publisher = self.create_publisher(Collision, Topics.MAPPED_COLLISIONS, 1)
+		self.right_ef_publisher = self.create_publisher(PoseStamped, Topics.MAPPED_OP_REF_POSE, 1)
+		self.left_ef_publisher = self.create_publisher(PoseStamped, Topics.MAPPED_OP_LEF_POSE, 1)
+
+		# Define service
+		self.operator_initial_head_pose = self.create_service(GetLinkPose, Services.INITIAL_HEAD_POSE, self.operator_initial_head_pose_callback)
+	
+	def get_roboy_link_name(self, operator_link):
+		"""Gets corresponding roboy link's name to the given operator link's name.
+		
+		Args:
+			operator_link (string): Operator link's name.
+
+		Returns:
+		   	string: Corresponding roboy link's name.
+
+		"""
+		roboy_link = None
+		for roboy, oper in zip(self.roboy_link_names, self.operator_link_names):
+			if oper == operator_link:
+				roboy_link = roboy
+				break
+		return roboy_link
+
+	def get_operator_link_name(self, roboy_link):
+		"""Gets corresponding operator link's name to the given roboy link's name.
+		
+		Args:
+			roboy_link (string): Roboy link's name.
+
+		Returns:
+		   	string: Corresponding operator link's name.
+
+		"""
+		operator_link = None
+		for roboy, oper in zip(self.roboy_link_names, self.operator_link_names):
```

Comment:
this assumes the same order of the links, which is not ideal.

Reply:
This mapping is defined in the params file and we cannot use dictionaries there, so we need two lists. Thats why I specified the importance of the lists order:
https://github.com/Roboy/bulletroboy/blob/833c8775de086dab569622e48959147a60f4aa22/config/conf.yaml#L3